File: src/backend/common-dto/src/main/java/dto/ApiResponse.java
package dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import dto.ErrorResponse;


@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private boolean success;
    private T data;
    private ErrorResponse error;

    // Public no-argument constructor for Jackson deserialization
    public ApiResponse() {
        this.success = false;
    }

    private ApiResponse(boolean success) {
        this.success = success;
    }


    public static <T> ApiResponse<T> success(T data) {
        ApiResponse<T> response = new ApiResponse<>(true);
        response.setData(data);
        return response;
    }

    public static <T> ApiResponse<T> error(int status, String message) {
        ApiResponse<T> response = new ApiResponse<>(false);
        response.setError(new ErrorResponse(status, message));
        return response;
    }

    public boolean isSuccess() {
        return success;
    }

    public void setSuccess(boolean success) {
        this.success = success;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    public ErrorResponse getError() {
        return error;
    }

    public void setError(ErrorResponse error) {
        this.error = error;
    }
}

File: src/swift-ui/AdOnWheelsApp/AdOnWheelsApp/Core/DTO/ApiResponse.swift
import Foundation

struct ApiResponse<T: Decodable>: Decodable {
    let success: Bool
    let data: T?
    let error: ErrorResponse?
}

struct ErrorResponse: Decodable, Error {
    let internalCode: Int
    let message: String
    let validationErrors: [String: String]?
    
    enum CodingKeys: String, CodingKey {
        case internalCode = "internalCode"
        case message = "message"
        case validationErrors = "validationErrors"
    }
}
